DRA: monitor plugin connection

bart0sh opened the pull request:
#### What type of PR is this?

/kind bug
/kind feature

#### What this PR does / why we need it:

It implements gRPC connection monitoring for DRA plugins to enable automatic deregistration and resource cleanup.

This is an alternative implementation of connection monitoring. [Previous implementation](https://github.com/kubernetes/kubernetes/pull/129799) raised certain concerns that should be addressed in this PR.

Unlike the previous implementation, this one:
- doesn't use GetInfo requests to monitor plugin connection. It uses native gRPC connection monitoring capabilities instead.
- has much smaller impact area as it monitors only DRA connections. Previous implementation impacted all 3 plugin types: CSI, Device and DRA. 
- monitors service connection. Previous implementation monitored registration connection.
- doesn't run any monitoring goroutines. Previous implementation run monitoring goroutine for every registered plugin.

#### Which issue(s) this PR fixes:

Fixes #128696
Ref #127457

#### Special notes for your reviewer:

This PR includes unit test fix from #131065. It will be removed when #131065 is merged.

#### Does this PR introduce a user-facing change?

```release-note
Kubelet unregisters DRA plugins and removes resource slices on connection drops.
```

/sig node
/cc @SergeyKanzhelev @ffromani @lauralorenz @pohly @swatisehgal

bart0sh committed:
kubelet: dra: test connection handling

Add TestConnectionHandling unit test to verify that ResourceSlices
are properly cleaned up when the DRA plugin's gRPC connection is
dropped, and retained when the connection is quickly restored.

bart0sh committed:
kubelet: dra: handle plugin connection

Detect plugin connection drops using a gRPC stats handler and trigger
asynchronous cleanup of ResourceSlices for the disconnected plugin.
If the plugin reconnects before cleanup completes, cancel the pending
cleanup to avoid removing resources for an active plugin. This ensures
stale resources are cleaned up after disconnects, while preventing
premature deletion on reconnects.

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-kind-dra-all
/test pull-kubernetes-kind-dra

bart0sh commented:
/retest

bart0sh commented:
/retest

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-kind-dra-all
/test pull-kubernetes-kind-dra

bart0sh commented:
/retest

bart0sh commented:
@pohly @ffromani @swatisehgal @SergeyKanzhelev PTAL, thanks!

jackfrancis commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -217,13 +217,15 @@ func (h *RegistrationHandler) RegisterPlugin(pluginName string, endpoint string,
+		conn:                nil,
+		endpoint:            endpoint,
+		chosenService:       chosenService,
+		clientCallTimeout:   timeout,
```
Looks like we're not making any changes to the timeout in this PR, but I wasn't able to confirm from a cursory source code review whether or not this timeout configuration is exposed in user space. Are users able to declare an opinionated timeout value here?

bart0sh replied:
It looks like clientCallTimeout is not used anywhere. It's passed to registration handler only from one place: https://github.com/kubernetes/kubernetes/blob/v1.33.0/pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go#L126 and its value is nil, which means that each plugin sets its predefined timeout. DRA sets it to 45 sec, CSI - to 2 min, Device plugin ignores passed value and always sets it to 10 seconds.

nojnhuh commented:
Could we document this new behavior in https://github.com/kubernetes/kubernetes/blob/e3b9541b235e47e7afd7edef403e1187f1e62a16/pkg/kubelet/pluginmanager/pluginwatcher/README.md?

In particular, some guidance around using livenessProbes to restart unregistered plugins would be helpful for https://github.com/kubernetes-sigs/dra-example-driver/issues/96.

bart0sh commented:
> Could we document this new behavior in https://github.com/kubernetes/kubernetes/blob/e3b9541b235e47e7afd7edef403e1187f1e62a16/pkg/kubelet/pluginmanager/pluginwatcher/README.md?

As this is DRA specific, we should probably document it somewhere in the DRA docs. Not sure if we have it, though.

bart0sh commented on pkg/kubelet/cm/dra/plugin/registration_test.go:
```diff
@@ -226,10 +243,65 @@ func TestRegistrationHandler(t *testing.T) {
+	_, err = client.ResourceV1beta1().ResourceSlices().Create(tCtx, slice, metav1.CreateOptions{})
+	require.NoError(t, err, "recreate slice")
+
+	// Simulate disconnect
+	require.NoError(t, conn.Close())
```
I'm using e2e test driver to simulate this kind of things. It's not easy, but doable. As a result you may have a nice new e2e test to cover a use case you're simulating.

pohly commented:
> As this is DRA specific, we should probably document it somewhere in the DRA docs. Not sure if we have it, though.

[pluginwatcher/README.md](https://github.com/kubernetes/kubernetes/blob/e3b9541b235e47e7afd7edef403e1187f1e62a16/pkg/kubelet/pluginmanager/pluginwatcher/README.md) already has documentation of seamless upgrades although that is also only supported for DRA.

I think a new section on "monitor plugin connection" makes sense,. Just add the same note that it is only implemented by DRA plugins.

bart0sh committed:
Plugin manager: documentation update

Added a description of the DRA connection monitoring to the
Kubelet plugin manager documentation.

bart0sh commented:
@pohly 
> I think a new section on "monitor plugin connection" makes sense,. Just add the same note that it is only implemented by DRA plugins.

Done, PTAL

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -30,6 +30,13 @@ setting `maxSurge` to a value larger than zero enables such a seamless upgrade.
 
+## Monitoring plugin connection
+
+If plugin drops GRPC connection with Kubelet, it gets automatically unregistered
```
GRPC -> gRPC

Can you add some more meat to this section? How does it work technically? What timeouts are there?

This document is mostly for developers, not users.

bart0sh replied:
Added, PTAL.

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -22,6 +22,16 @@ should end with a DNS domain that is unique for the plugin. Each time a plugin
 
+## Monitoring plugin connection
+
+Kubelet monitors gRPC connection to a plugin using stats connection monitoring
+handler https://github.com/grpc/grpc-go/blob/master/examples/features/stats_monitoring/README.md
+that allows to hook on connection start and stop events. If plugin drops
+connection with Kubelet, it gets automatically unregistered and its resources
+are cleaned up.
```
```suggestion
The kubelet monitors the gRPC connection to a plugin's
registration socket using a [stats connection monitoring
handler](https://github.com/grpc/grpc-go/blob/master/examples/features/stats_monitoring/README.md).
Such a handler can observe connection start and stop events.
If a plugin drops the connection with the kubelet, it gets 
automatically unregistered immediately and its resources
are cleaned up. This works because the connection is
using a reliable Unix Domain stream, so loosing the connection
is guaranteed to mean that the peer closes its side of
the connection and not some intermittent network timeout.
```

Is this a correct update?

What about reconnecting to a socket? I can imagine why it's not needed (the plugin cannot reopen the socket without recreating it), but this should be spelled out explicitly as one of the assumptions behind the implementation.

pohly replied:
Also, why is monitoring the registration socket sufficient?

pohly replied:
This is absolutely worth calling out here, together with the rationale for doing it this way.

pohly replied:
I also expect that it could have implications for retries:
- The registration socket might be provided by one container, the service socket by another (ala CSI sidecar deployment).
- Service container restarts, dropping the connect.
- It restarts.
- Does the kubelet re-register the plugin?

bart0sh replied:
It should, because plugin manager monitors socket directory for new entries. When Service container restarts, the socket is closed and deleted and then it's re-created again on the container start.

pohly replied:
It seems we are not using the same terminology. Looking at the README.md, "registration socket" and "service socket" are not defined. What I meant is "registration socket" = where the plugin provides the [pluginregistration](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubelet/pkg/apis/pluginregistration)/[v1](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubelet/pkg/apis/pluginregistration/v1) `Registration` service and "service socket" = where the plugin provides the DRA service. You seem to use the opposite definition.

Looks like the README text will need a few more words than "registration socket" to describe what it is monitoring...

bart0sh replied:
> Also, why is monitoring the registration socket sufficient?

Registration socket is re-connected often as far as I remember, so it's not possible to use this approach for it without changing plugin manager, which I was trying to avoid.

bart0sh replied:
> What I meant is "registration socket" = where the plugin provides the [pluginregistration](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubelet/pkg/apis/pluginregistration)/[v1](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubelet/pkg/apis/pluginregistration/v1) Registration service and "service socket" = where the plugin provides the DRA service. You seem to use the opposite definition.

I'm using the same terminology. Connection monitoring is setup in the pkg/kubelet/cm/dra/plugin/plugin.go and uses to serve service connection (NodePrepareResources, NodeUnprepareResources gRPC APIs).

pohly replied:
Then I don't understand what you mean with "plugin manager monitors socket directory for new entries".

It monitors `/var/lib/kubelet/plugins_registry`. Where does it monitor the directory that contains the endpoint returned by the plugin? Is that new code that you are adding?

pohly replied:
> Registration socket is re-connected often as far as I remember, so it's not possible to use this approach for it without changing plugin manager, which I was trying to avoid.

Not changing the plugin manager makes sense. But that then raises another question: connections to the DRA driver are established when needed and then cached, right?

What if the node is currently idle? Will the kubelet detect that a registered DRA driver isn't really ready to serve requests?

Instead of speculating, would it make sense to do test driven development? You could write tests for the interesting scenarios and then verify that they are handled by the existing code. If not, extend the code.

I can see at least three test scenarios where registration and DRA (= service socket) are different:
- DRA driver comes up, the kubelet connects to the DRA socket, then the DRA driver disconnects the DRA socket.
- DRA driver comes up, registers its DRA socket, but then fails to to listen on that socket (could happen in a sidecar scenario).
- DRA driver comes up, the kubelet connects to the DRA socket, then the plugin restarts and recreates only the DRA socket.

In all scenarios we want the kubelet to treat the DRA driver as not registered immediately, which starts the grace period before it removes ResourceSlices. In the last scenario we want the kubelet to detect that the driver is back and not remove ResourceSlices, as long as it comes back quickly enough.

Similar scenarios minus the second one are possible when registration and DRA socket are identical. This also needs to be tested.

bart0sh replied:
> I also expect that it could have implications for retries:
> 
> * The registration socket might be provided by one container, the service socket by another (ala CSI sidecar deployment).
> * Service container restarts, dropping the connection.
> * It restarts.
> * Does the kubelet re-register the plugin?

According to the https://github.com/kubernetes/kubernetes/pull/128907 it doesn't as registration socket doesn't change, but that's another story. I'm going to address it separately, hopefully by helping to merge proposed solution.

bart0sh replied:
> What if the node is currently idle? Will the kubelet detect that a registered DRA driver isn't really ready to serve requests?

In this case DRA(service) connection is not established yet, correct? Then there is nothing to monitor. So, I'd say this is another case, which can be addressed separately, modifying the plugin manager, for example we can reopen https://github.com/kubernetes/kubernetes/pull/129799 to address this use case.

bart0sh replied:
> DRA driver comes up, the kubelet connects to the DRA socket, then the DRA driver disconnects the DRA socket.
> DRA driver comes up, the kubelet connects to the DRA socket, then the plugin restarts and recreates only the DRA socket.

These 2 cases should be handled by this implementation. I'll try to implement e2e_node tests for them if that's needed.

bart0sh replied:
> DRA driver comes up, registers its DRA socket, but then fails to to listen on that socket (could happen in a sidecar scenario).

This one should be handled separately in my opinion, either by merging https://github.com/kubernetes/kubernetes/pull/128907 or https://github.com/kubernetes/kubernetes/pull/129799.

bart0sh replied:
In short - instead of addressing all possible issues in one PR, I'd prefer to fix them step by step. Monitoring DRA service connection is a first step.

bart0sh replied:
> Let's first extend the README.md with a description of what the code is expected to handle, with an explanation of how, and what it does not handle, again with why. As it stands now, we had to have a long debate over what it really solves.

OK. As most of the implementation is in the DRA plugin code, do you still think this README.md is the right place for its documentation?
I'd rather put it into pkg/kubelet/cm/dra/README.md or pkg/kubelet/cm/dra/plugin/README.md.

pohly replied:
Yes, I think it's the right place. It's better to have all information about kubelet plugins in one place.

bart0sh replied:
Updated the README.md again. PTAL.

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -22,6 +22,16 @@ should end with a DNS domain that is unique for the plugin. Each time a plugin
+connection with Kubelet, it gets automatically unregistered and its resources
+are cleaned up.
+
+**Warning**: Monitoring plugin connection is only supported for DRA at the moment.
```
```suggestion
**Warning**: Monitoring the plugin connection is only supported for DRA at the moment.
```

bart0sh commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -22,6 +22,44 @@ should end with a DNS domain that is unique for the plugin. Each time a plugin
+   observes a `ConnEnd` event.
+4. This immediately triggers plugin deregistration and removal from the plugin
+   manager’s desired state of the world.
+5. The service socket file is deleted when the plugin terminates, making
```
> Deleted by whom? The plugin might skip that step.

By the unix socket listener as far as I remember.

pohly replied:
If the process is killed, that won't happen. Let's remove the assumption that the service socket file gets deleted, because it is not guaranteed.

bart0sh commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -22,6 +22,44 @@ should end with a DNS domain that is unique for the plugin. Each time a plugin
+4. This immediately triggers plugin deregistration and removal from the plugin
+   manager’s desired state of the world.
+5. The service socket file is deleted when the plugin terminates, making
+   reconnection impossible unless the plugin re-registers.
```
> Can't the plugin remove the old service socket and start listening again, without ever doing anything with the registration socket?

I don't think this scenario is realistic, considering that plugins usually call `grpc.Server.Serve` in a go-routine and its failure to serve is considered a fatal error even if processed at all. However, re-registering plugin on `ConnStart` shouldn't be hard to implement, so added to the doc. Thanks for pointing out!

bart0sh replied:
Anything else is still missing in the doc? If not, can you outline e2e tests that need to be implemented?
Here is what I'd test:
- Plugin disconnects and then restarts. It's expected to be deregistered and then registered again.
- Plugin disconnects and then reconnects on the same socket. It's expected to be deregistered and then registered again.
- Plugin registers. The service connection is expected to be established even without any service gRPC calls.

bart0sh replied:
@pohly 
> Can't the plugin remove the old service socket and start listening again, without ever doing anything with the registration socket?

> My expectation is that gRPC would move the connection back to "connected" after some time. It might only happen if there is traffic pending.

I guess it would be another connection and I doubt this scenario is supported by the current code.
Anyway, I tried to simulate this scenario by stopping only DRA gRPC service(tried also closing underlying socket without stopping the service) and keeping registration service running, but I can only see that the old connection ends, but don't see any signs of connection (re)start. I tried this experiment with and without removing a socket file. See 2 latest commits for the implementation details.

This could probably be supported by re-establishing connection from the client (Kubelet) side after the connection drop, but this is another story as far as I understand. We're talking about internal gRPC reconnection mechanisms here, correct? Those either don't exist or don't work in the scenario when unix socket listener closes and opens again on the same socket path. At least this is what I can see with my experiments.

pohly replied:
What if the stats handler triggers a reconnect? It could either call https://pkg.go.dev/google.golang.org/grpc#ClientConn.Connect (experimental) or call some gRPC method (which one TBD).

bart0sh committed:
kubelet: dra: connect to plugin at registration

Establish a gRPC connection to the DRA plugin immediately during
registration instead of waiting for the first gRPC call. This change
improves connection monitoring and prevents inconsistent behavior if
the connection drops between registration and first use. The connection
is triggered asynchronously and does not block registration.

Switch from the deprecated grpc.Dial to grpc.NewClient for creating
the gRPC connection. grpc.NewClient is the recommended API and provides
a more modern and robust way to establish client connections. The code
now calls Connect() on the client to initiate the connection process
without blocking.

bart0sh commented:
/retest

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -22,6 +22,46 @@ should end with a DNS domain that is unique for the plugin. Each time a plugin
+- Deregister the plugin immediately when the connection is lost
+- Trigger cleanup of the plugin’s resources
+
+**registration socket** is monitored by the plugin manager. It registers the
+plugin when registration socket is created by the plugin and deregisters the
+plugin when socket is removed. After registration, the plugin begins serving
+gRPC requests over the **service socket**. Monitoring this service socket is
+therefore more accurate for detecting the real availability of the plugin.
```
```suggestion
The **registration socket** is monitored by the plugin manager. It registers the
plugin when registration socket is created by the plugin and
the GetInfo gRPC call succeeds.
It  deregisters the plugin when the socket is removed.

The plugin should be ready to handle
gRPC requests over the **service socket** that
it returned in response to the GetInfo call because
the kubelet might try to use the service immediately.
Monitoring this service socket is
therefore more accurate for detecting the real availability of the plugin.
```

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -22,6 +22,46 @@ should end with a DNS domain that is unique for the plugin. Each time a plugin
+   observes a `ConnEnd` event.
+4. This immediately triggers plugin deregistration and removal from the plugin
+   manager’s desired state of the world.
+5. If the plugin connects again, the stats handler observes a `ConnBegin` event
```
```suggestion
5. If the plugin starts listening again, gRPC automatically reconnects, the stats handler observes a `ConnBegin` event
```

bart0sh replied:
gRPC doesn't automatically reconnects. Kubelet has to do it explicitly.

pohly replied:
I'm not convinced that this is true. In general, whether a gRPC connection is established is transparent to the application layer. It issues gRPC requests and gRPC ensures that they get to the server, (re)connecting if necessary.

pohly replied:
Do you want to write a stand-alone test for this assumption or shall I?

bart0sh replied:
Yes, I'll do it.

bart0sh replied:
here you go:

server.go
```
package main

import (
	"log"
	"net"
	"os"
	"time"

	"google.golang.org/grpc"
)

func main() {
	socketPath := "/tmp/grpc.sock"
	// Remove socket file if it exists
	if err := os.RemoveAll(socketPath); err != nil {
		log.Fatalf("failed to remove existing socket: %v", err)
	}
	for {
		log.Println("Starting gRPC server...")

		lis, err := net.Listen("unix", socketPath)
		if err != nil {
			log.Fatalf("failed to listen on unix socket: %v", err)
		}

		s := grpc.NewServer()

		go func() {
			if err := s.Serve(lis); err != nil {
				log.Printf("server exited: %w\n", err)
			}
		}()

		time.Sleep(20 * time.Second)
		log.Println("Stopping gRPC server...")
		s.GracefulStop()
		time.Sleep(5 * time.Second)
	}
}

```

client.go
```
package main

import (
        "log"
        "time"

        "google.golang.org/grpc"
        "google.golang.org/grpc/stats"
        "golang.org/x/net/context"
)

const socketPath = "unix:///tmp/grpc.sock"

type connStatsHandler struct{}

func (h *connStatsHandler) TagConn(ctx context.Context, info *stats.ConnTagInfo) context.Context {
        return ctx
}

func (h *connStatsHandler) HandleConn(ctx context.Context, s stats.ConnStats) {
        switch s.(type) {
        case *stats.ConnBegin:
                log.Println("[StatsHandler] Connection began")
        case *stats.ConnEnd:
                log.Println("[StatsHandler] Connection ended")
        }
}

func (h *connStatsHandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) context.Context {
        return ctx
}

func (h *connStatsHandler) HandleRPC(ctx context.Context, s stats.RPCStats) {}

func main() {
        conn, err := grpc.Dial(
                socketPath,
                grpc.WithInsecure(),
                grpc.WithBlock(),
                grpc.WithStatsHandler(&connStatsHandler{}),
                grpc.WithBackoffMaxDelay(5*time.Second),
        )
        if err != nil {
                log.Fatalf("failed to connect: %v", err)
        }
        defer conn.Close()

        log.Println("Client is connected. Waiting indefinitely.")
        select {}
}
```

Here is what I see on the server console:
```
$ go run server.go
2025/05/20 18:40:05 Starting gRPC server...
2025/05/20 18:40:25 Stopping gRPC server...
2025/05/20 18:40:30 Starting gRPC server...
2025/05/20 18:40:50 Stopping gRPC server...
2025/05/20 18:40:55 Starting gRPC server...
2025/05/20 18:41:15 Stopping gRPC server...
2025/05/20 18:41:20 Starting gRPC server...
2025/05/20 18:41:40 Stopping gRPC server...
2025/05/20 18:41:45 Starting gRPC server...
2025/05/20 18:42:05 Stopping gRPC server...
2025/05/20 18:42:10 Starting gRPC server...
```

and on the client console:
```
$ go run client.go
2025/05/20 18:40:12 [StatsHandler] Connection began
2025/05/20 18:40:12 Client is connected. Waiting indefinitely.
2025/05/20 18:40:25 [StatsHandler] Connection ended
```

bart0sh replied:
It works only when there is active RPC traffic. I can see reconnections after adding a simple Ping service:

```
$ go run client.go
2025/05/20 19:49:40 [StatsHandler] Connection began
2025/05/20 19:49:40 Client is connected. Sending Ping every 3 seconds...
2025/05/20 19:49:43 Ping OK
2025/05/20 19:49:46 Ping OK
2025/05/20 19:49:49 Ping OK
2025/05/20 19:49:52 Ping OK
2025/05/20 19:49:54 [StatsHandler] Connection ended
2025/05/20 19:49:55 Ping failed: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial unix /tmp/grpc.sock: connect: no such file or directory"
2025/05/20 19:49:58 Ping failed: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial unix /tmp/grpc.sock: connect: no such file or directory"
2025/05/20 19:50:01 [StatsHandler] Connection began
2025/05/20 19:50:01 Ping OK
2025/05/20 19:50:04 Ping OK
2025/05/20 19:50:07 Ping OK
```

bart0sh replied:
In practice it means that we can only detect that the plugin is back online when we send [Un]Prepare requests to it, which is not exactly what we want, but probably better than nothing.

pohly replied:
Or use Connect (https://github.com/kubernetes/kubernetes/pull/131073#discussion_r2097849217).

Connect feels more appropriate for our purposes, the only drawback is that it is marked "Experimental". I think I would be okay with taking that risk.

bart0sh replied:
I'd simply re-dial by calling [getOrCreateGRPCConn](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/dra/plugin/plugin.go#L67). It looks heavier, but it'll be called only after connection drop, which shouldn't happen too often.

pohly replied:
I think that's worse than using `Connect`:
- As it stands, it uses `grpc.Dial`, which is deprecated and perhaps worse than using an experimental API.
- It times out, whereas what we need is trying to connect forever until the connection really isn't needed anymore.
- It blocks until it times out, so you need to run it in a goroutine.

This actually brings me to another interesting question: at what point should the DRA manager stop trying to connect (no matter how it does that)?

Normally the answer would be "until it is told by the plugin manager that the plugin is not registered anymore" - but the DRA manager itself unregisters the plugin now, so it cannot use that as a signal.

Perhaps the DRA manager shouldn't unregister the plugin on a service connection loss? We then would have more states:
- A plugin is registered or not (tracked by plugin manager).
- While it is registered, it is usable or not (tracked by DRA manager).

We can trigger cleanup of ResourceSlices when it a plugin becomes unregistered or unusable and abort the cleanup when it becomes usable again. That should go a long way towards solving the problem that this PR is trying to address. The "deregistration" part is less important.

pohly replied:
> I'd introduce a timeout.

And then what? We have to keep monitoring and retrying because that is what the kubelet did before.

bart0sh replied:
I was thinking about unregistering plugin completely, but you're right, we need to keep retrying unless we decide to break compatibility with the previous behaviour.

bart0sh replied:
> Will try to keep connection monitoring in the DRA code.

Done. PTAL. e2e needs to be done still, but it would be great if you can review the implementation and the documentation.

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -22,6 +22,46 @@ should end with a DNS domain that is unique for the plugin. Each time a plugin
+5. If the plugin connects again, the stats handler observes a `ConnBegin` event
+   and adds the plugin to the plugin manager’s desired state of the world, which
+   triggers plugin re-registration.
+6. If the plugin restarts it must explicitly re-register with the Kubelet. This
```
Why? I thought we wanted to support the scenario where the driver consists of a permanently running "registration" side car and only the service container restarts.

pohly replied:
Let's assume that gRPC reconnects. Then this PR introduces a regression. Previously, a plugin was allowed to restart the service and the kubelet would continue talking to it when it comes back. Now we introduce "plugin must register again" as a new requirement which might not be implemented by all plugins.

pohly replied:
Do you agree that we need to handle this, even if it makes the implementation more complex?

bart0sh replied:
Yes, I agree with this. Will update the documention.

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -22,6 +22,46 @@ should end with a DNS domain that is unique for the plugin. Each time a plugin
+   ensures that the Kubelet only interacts with plugins that are fully initialized
+   and healthy.
+
+**Warning**: Monitoring the plugin connection is only supported for DRA at the moment.
```
Let's put this directly after "Monitoring Plugin Connection".

pohly reviewed (commented):
Another pass over the README.md to understand what to expect from the code. One question about "must explicitly re-register"...

bart0sh commented:
/retest

bart0sh commented:
/test pull-kubernetes-kind-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented:
/retest

bart0sh commented on pkg/kubelet/cm/dra/plugin/plugin.go:
```diff
@@ -179,3 +188,32 @@ func newMetricsInterceptor(pluginName string) grpc.UnaryClientInterceptor {
+		logger.V(2).Info("Connection end", "plugin", p.name, "endpoint", p.endpoint, "stats", connStats)
+		p.connectionDropped = true
+		if p.registrationHandler != nil {
+			p.registrationHandler.cleanupResourceSlices(p.name)
```
@pohly e2e tests are flaking probably because this call conflicts with the one called when plugin is un-registered. I'm looking at it, but your help would be appreciated as you know that code better.

pohly replied:
See https://github.com/kubernetes/kubernetes/pull/131073/files#r2104086671. I think state tracking is simply incomplete at the moment.

bart0sh commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -287,52 +300,63 @@ func (h *RegistrationHandler) DeRegisterPlugin(pluginName, endpoint string) {
+	// (see RegisterPlugin).
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+	if cancel := h.pendingWipes[pluginName]; cancel != nil {
+		(*cancel)(errors.New("plugin deregistered a second time"))
+	}
+	h.pendingWipes[pluginName] = &cancel
```
@pohly I suspect this code shouldn't cancel ongoing wiping, but let it continue. How about doing something like this at the beginning of this function?
```
if cancel := h.pendingWipes[pluginName]; cancel != nil {
        logger.V(3).Info("Previously started cleanup detected, do nothing", "pluginName", pluginName)
	return
}
```

bart0sh commented:
/test pull-kubernetes-kind-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

pohly commented on pkg/kubelet/cm/dra/plugin/plugin.go:
```diff
@@ -179,3 +188,32 @@ func newMetricsInterceptor(pluginName string) grpc.UnaryClientInterceptor {
+		if p.connectionDropped {
+			if p.registrationHandler != nil {
+				logger.V(2).Info("Stopping ResourceSlices cleanup", "plugin", p.name, "endpoint", p.endpoint)
+				p.registrationHandler.Stop()
```
Why is this calling https://github.com/kubernetes/kubernetes/blob/f0a7d04a5b8382a8b6f92ab4f5c41125405c3add/pkg/kubelet/cm/dra/plugin/registration.go#L115-L119?

That shuts down the entire registration handler for all drivers, which is not what we want here.

What the registration handler needs to track is which service sockets are registered and whether they are usable. The second part implies adding new APIs. The implementation then needs to take care of starting wiping when the number of usable connections drops to zero and canceling that when it becomes non-zero again.

Perhaps triggering wiping shouldn't be in `registration.go` at all. With what I said above, `plugins_store.go` might be a better place. Then we can also change https://github.com/kubernetes/kubernetes/blob/f0a7d04a5b8382a8b6f92ab4f5c41125405c3add/pkg/kubelet/cm/dra/plugin/plugins_store.go#L47-L50 so that it only considers usable connections.

bart0sh replied:
> That shuts down the entire registration handler for all drivers, which is not what we want here.

My bad. I was going to cancel pending wiping at this point.

> What the registration handler needs to track is which service sockets are registered and whether they are usable. The second part implies adding new APIs. The implementation then needs to take care of starting wiping when the number of usable connections drops to zero and canceling that when it becomes non-zero again.

I don't fully understand this. I thought that it's enough to start wiping when the plugin connection drops and cancel 
ongoing wiping when the connection is restored. Can you elaborate?

bart0sh replied:
That makes sense, thanks! Is there anything that can help to distinguish plugins or their connections? Any attribute, etc? If this exists we can use pluginName + this identifier to store wiping info. It can help canceling a correct wiping process.

bart0sh replied:
nvm, this is probably not needed. I just need to count usable connections as you've explained, thanks!

bart0sh replied:
> Then we can also change
[kubernetes/pkg/kubelet/cm/dra/plugin/plugins_store.go#L47-L50](https://github.com/kubernetes/kubernetes/blob/f0a7d04a5b8382a8b6f92ab4f5c41125405c3add/pkg/kubelet/cm/dra/plugin/plugins_store.go#L47-L50) so that it only considers usable connections.

This could cause PrepareResources and UnprepareResources to be called on 2 different plugins if the last one is disconnected. Is this ok to do?

pohly replied:
Yes. The docs for seamless upgrades explicitly call this out.

pohly replied:
https://github.com/kubernetes/kubernetes/blob/122219790f82f5843d99aacbcb3d96162161e027/staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go#L277-L285

bart0sh replied:
> Perhaps triggering wiping shouldn't be in registration.go at all. With what I said above, plugins_store.go might be a better place. Then we can also change

I hesitated to move it to plugin_store.go to avoid making big code changes. The rest you've suggested should be implemented now. CI is green. PTAL.

bart0sh replied:
I was too fast. Unit tests are failing due to changed plugin store `get` API. fixing now. I'll change this PR status to the "Ready for review" when I'm done and let you know.

pohly commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -217,15 +217,28 @@ func (h *RegistrationHandler) RegisterPlugin(pluginName string, endpoint string,
+	// If this is not done, the connection will be created lazily when
+	// the first gRPC call is made, which can lead to inconsistent behavior
+	// if the connection drops between registration and the first gRPC call.
+	conn, err := pluginInstance.getOrCreateGRPCConn()
```
This causes a new failure mode: previously, a plugin could register a service socket before listening on it. It shouldn't, but it could. Then attempts to use the service fail and get retried. Now registration fails, with unknown consequences. Is the plugin told that registration failed? Does it then shut down, causing a restart loop?

This needs to be another E2E test:
- Listen on the registration socket first.
- Wait for successful registration.
- Listen on the service socket.
- Ensure that NodePrepareResources succeeds.

To achieve this, I think we should remove the connection timeout in `getOrCreateGRPCConn`. We don't really care how long it takes to actually establish the connection. We still have the timeouts for individual gRPC method calls.

This leads to another corner case (and E2E test case):
- Listen on the registration socket first.
- Wait for successful registration.
- *Don't* listen on the service socket.
- Expect that resource slices get wiped after the usual timeout.

This might follow from how usable connections are tracked. Remains to be seen.

bart0sh replied:
> This causes a new failure mode: previously, a plugin could register a service socket before listening on it.

Correct. Moreover, this is a regression. I'll change the code to run `pluginInstance.getOrCreateGRPCConn()` in a go-routine, so registration wouldn't depend on it.

> This might follow from how usable connections are tracked.

I don't understand the whole idea of usable connections to be honest. So far I thought that we're considering all connection usable and all connection drops temporary.

pohly replied:
"usable right now" is what we detect and react to. That changes from true to false and potentially back.

pohly replied:
We can call it "a connection is connected", but that is a bit weird.

bart0sh replied:
Current implementation uses grpc.NewClient to initiate connection. This call is non-blocking, so it should work just fine. 
I've implemented 2 e2e test cases you've mentioned above.

pohly commented:
@bart0sh: can you convert the PR to a draft and selectively invoke jobs through `/test` until it is ready?

bart0sh commented:
/test pull-kubernetes-kind-dra

bart0sh commented:
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented:
/test pull-kubernetes-kind-dra-all

bart0sh committed:
kubelet: dra: update unit tests

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd-2-0-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-kind-dra-all

bart0sh committed:
e2e: dra: add ability to start only registrar or plugin

This change introduces StartOnlyPlugin and StartOnlyRegistrar functions,
allowing tests and deployments to start just the plugin service or just
the registrar. This enables more flexible testing and deployment
scenarios, such as verifying registrar or plugin behavior in isolation.

bart0sh committed:
e2e_node: dra: implement reliability test cases

Add 3 reliability tests for the DRA plugin in e2e_node:
- ensure that the plugin is functional when it starts listening
  on its service socket after registration.
- verify that ResourceSlices are removed if the plugin stops after
  registration.
- verify that ResourceSlices are removed if the plugin doesn't
  get connected after registration.

bart0sh commented:
/test pull-kubernetes-verify
/test pull-kubernetes-unit
/test pull-kubernetes-typecheck

bart0sh commented:
/test pull-kubernetes-verify
/test pull-kubernetes-unit
/test pull-kubernetes-typecheck

/test pull-kubernetes-kind-dra
/test pull-kubernetes-kind-dra-all
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented on pkg/kubelet/cm/dra/plugin/plugin.go:
```diff
@@ -77,34 +96,48 @@ func (p *Plugin) getOrCreateGRPCConn() (*grpc.ClientConn, error) {
+	conn, err := grpc.NewClient(network+"://"+p.endpoint,
 		grpc.WithTransportCredentials(insecure.NewCredentials()),
 		grpc.WithContextDialer(func(ctx context.Context, target string) (net.Conn, error) {
+			target = strings.Split(target, "//")[1] // Remove the network prefix
```
It turned out that the NewClient works just fine with a default dialer, so I removed it. We don't have a connect timeout anymore, so this seemed reasonable to me.

bart0sh commented on pkg/kubelet/cm/dra/plugin/plugin.go:
```diff
@@ -77,34 +96,48 @@ func (p *Plugin) getOrCreateGRPCConn() (*grpc.ClientConn, error) {
+// WaitForConnection repeatedly attempts to establish a connection to a DRA
+// plugin within the specified timeout. It polls at the given interval until
+// a connection is established or the timeout is reached. The function returns
+// an error if a connection cannot be established within the timeout period.
```
Thanks for pointing out, moved to testing.go

bart0sh commented on pkg/kubelet/cm/dra/plugin/plugin.go:
```diff
@@ -179,3 +212,29 @@ func newMetricsInterceptor(pluginName string) grpc.UnaryClientInterceptor {
+		logger.V(2).Info("Connection begin", "plugin", p.name, "endpoint", p.endpoint, "stats", connStats)
+		p.connected = true
+		if p.registrationHandler != nil {
+			p.registrationHandler.cancelCleanup(p.name, p.cancelCleanup)
```
> Are we sure here that the connection is still registered?

Added a check for that. We don't need to do any connection monitoring if plugin is not registered.

> I stand by what I said earlier: the correctness of the code would be easier to see and verify if cleanup handling was started and stopped in the plugin cache, with proper locking and checking of the current state.

That may make sense and I'm trying to implement it in a separate branch locally. However, I'm a bit worried that it will increase amount of code changes. Can we do this refactoring in a separate PR?

bart0sh commented:
/test pull-kubernetes-kind-dra
/test pull-kubernetes-kind-dra-all
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented:
/test pull-kubernetes-verify
/test pull-kubernetes-unit
/test pull-kubernetes-typecheck

bart0sh commented:
/test pull-kubernetes-verify
/test pull-kubernetes-unit
/test pull-kubernetes-typecheck

/test pull-kubernetes-kind-dra
/test pull-kubernetes-kind-dra-all
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented:
/retest

bart0sh commented:
/test pull-kubernetes-verify
/test pull-kubernetes-unit
/test pull-kubernetes-typecheck

/test pull-kubernetes-kind-dra
/test pull-kubernetes-kind-dra-all
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh committed:
don't run some tests on Windows

pohly commented on pkg/kubelet/cm/dra/plugin/plugin.go:
```diff
@@ -76,35 +89,65 @@ func (p *Plugin) getOrCreateGRPCConn() (*grpc.ClientConn, error) {
+			p.setCancelCleanup(p.registrationHandler.cleanupResourceSlices(p.name))
+		}
+	}
+}
```
Distributing the logic of when to start and stop cleanup among several different places makes the APIs more complex and harder to verify that all corner cases are handled correctly. For example, this looks like it triggers cleanup regardless whether some other socket is available for the driver, which probably explains the "rolling upgrade" failure.

I don't think we can merge this without the restructuring that I mentioned before.

bart0sh replied:
> Distributing the logic of when to start and stop cleanup among several different places makes the APIs more complex and harder to verify that all corner cases are handled correctly. For example, this looks like it triggers cleanup regardless whether some other socket is available for the driver, which probably explains the "rolling upgrade" failure.

Do I understand this correctly, that you want all cleanup triggering logic to be in one place, in the plugins store to be exact? Does this mean that connection monitoring should also be there? So far I found it logical to have it in the Plugin methods as Plugin has 1:1 relationship with connection. If you still think that we should only move cleanup machinery from registration handler to plugin store, then it will change nothing in connection monitoring, i.e. instead of triggering/cancelling cleanup through the registration handler, Plugin would need to do it through plugin store.

Speaking of plugin store. To handle cleanups it should be (re)initialized with `kubeClient` and `getNode`. This could be done by either keeping it a singleton object like it is now and initializing `kubeClient` and `getNode` on the start of the DRA Manager, or converting it into an usual object, that's created when DRA Manager starts. Which way would you recommend to use? Both will add more code changes to this PR, but the latter will probably add more.

BTW, I find it a bit surprising that you're suggesting that plugin store handles something else except of storing plugins, which is what it does currently. Wiping out resource slices looks out of scope of this structure, at least to me. Registration manager is also not ideal, but plugin store is even worse.

pohly replied:
> Do I understand this correctly, that you want all cleanup triggering logic to be in one place,

Yes.

> in the plugins store to be exact?

The plugin manager already knows about one half of information (endpoints). It just needs to be notified about connection changes. It also already has a mutex.

> converting it into an usual object, that's created when DRA Manager starts. 

That it's a global singleton has always felt wrong to me. I don't know why it was done this way. Testing is awfully complex because of that. If we can change it, that would be best.

> Wiping out resource slices looks out of scope of this structure, at least to me. Registration manager is also not ideal, but plugin store is even worse.

I see the plugin store as the central entity. The registration manager is just implementing a particular interface on top of it. I wouldn't mind having the plugin store implement that interface directly.

bart0sh committed:
kubelet: dra: move resource slice cleanup to cleanupHandler

This commit refactors the DRA plugin and registration handler
logic to move resource slice cleanup and pending wipe management
into a new cleanupHandler type. The plugin and registration
handler now delegate cleanup and cancellation responsibilities to
cleanupHandler. This improves separation of concerns, reduces
mutex contention, and makes the code easier to maintain and test.

bart0sh commented:
/test pull-kubernetes-verify
/test pull-kubernetes-unit
/test pull-kubernetes-typecheck

/test pull-kubernetes-kind-dra
/test pull-kubernetes-kind-dra-all
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented:
/test pull-kubernetes-verify
/test pull-kubernetes-unit
/test pull-kubernetes-typecheck

/test pull-kubernetes-kind-dra
/test pull-kubernetes-kind-dra-all
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

pohly commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -234,13 +127,29 @@ func (h *RegistrationHandler) RegisterPlugin(pluginName string, endpoint string,
+		// ResourceSlices. This will ensure that if the plugin never comes
+		// online, the ResourceSlices will be cleaned up after a delay.
+		logger.V(3).Info("Plugin is not connected, scheduling cleanup of ResourceSlices")
+		h.cleanupHandler.cleanupResourceSlices(pluginName)
```
At first glance this still looks wrong: what if there is some other socket which is currently connected?

I know that it just "looks" wrong when looking at this code, because you added more if checks inside `cleanupResourceSlices` (`if draPlugins.pluginConnected()`), but I just can't convince myself that all the different state transitions are handled correctly. Perhaps they are, perhaps they aren't. I can't tell.

This logs `Plugin is not connected, scheduling cleanup of ResourceSlices` in cases when nothing gets done. This will be confusing.

As it stands, the logic for starting and stopping wiping is scattered across three files and types (cleanup.go, registration.go, plugin.go), governed by several different mutexes. That's too much.

bart0sh replied:
> This logs Plugin is not connected, scheduling cleanup of ResourceSlices in cases when nothing gets done.

I agree, this is confusing. I'd remove them and make cleanup implementation responsible for logging.

pohly commented on pkg/kubelet/cm/dra/plugin/registration_test.go:
```diff
@@ -226,10 +266,87 @@ func TestRegistrationHandler(t *testing.T) {
+			} else {
+				// Simulate reconnect
+				plugin.closeConnection()
+				conn, err := plugin.getOrCreateGRPCConn()
```
This is not a realistic test. What we need to test is that the gRPC connection monitoring detects a restarted gRPC server.

I've copied this test over into https://github.com/kubernetes/kubernetes/pull/132058 and made that change there.

bart0sh replied:
> This is not a realistic test.

This test is testing connection drop. What's unrealistic in it?

pohly commented:
/close

Replaced by https://github.com/kubernetes/kubernetes/pull/132058.
